### DISCUSSION TITLE
Why the profiles table can not be updated?

### DISCUSSION THREAD

**Discussion by Albert-Gao** [supabase/supabase#15516] [Questions] [answered]
Hi, I have a profiles table that syncs from the `auth.user`, mostly uses the SQL copied from the doc,

The read operation has been disabled and I tested it works great,
However, I tried the update and it is not working at all...

I thought by the rule, a user should be able to update its own record, why it does not work?
 
What am I missing here?

Thanks :)

## The SQL
```sql
create table profiles
(
    id        uuid primary key,
    createdAt timestamp with time zone DEFAULT now(),
    updatedAt timestamp with time zone DEFAULT now(),

    nickName  varchar(256),
    avatarUrl varchar(5120),
    email     varchar(256),

    provider  varchar(256)
);



alter table profiles
    enable row level security;

create policy "profiles:disable read all" on profiles
    for select
    using (false);

create policy "profiles:update self" on profiles
    for update
    to authenticated, anon
    using (auth.uid() = id);


-- Set up Storage!
insert into storage.buckets (id, name)
values ('avatars', 'avatars');



-- Set up access controls for storage.
-- See https://supabase.com/docs/guides/storage#policy-examples for more details.
create policy "avatar-images:read all" on storage.objects
    for select using (bucket_id = 'avatars');

create policy "avatar-image:create self" on storage.objects
    for insert with check (bucket_id = 'avatars');



-- This trigger automatically creates a profile entry when a new user signs up via Supabase Auth.
-- See https://supabase.com/docs/guides/auth/managing-user-data#using-triggers for more details.
create function public.handle_new_user()
    returns trigger as
$$
begin
    insert into public.profiles (id, email, nickName, avatarUrl, createdAt, updatedAt)
    values (
        new.id, 
        new.email, 
        new.raw_user_meta_data ->> 'full_name', 
        new.raw_user_meta_data ->> 'avatar_url',
        new.created_at, 
        new.updated_at
    );
    return new;
end;
$$ language plpgsql security definer set search_path = public;
create trigger on_auth_user_created
    after insert on auth.users
    for each row
    execute procedure public.handle_new_user();
```

## The test
![image](https://github.com/supabase/supabase/assets/18282328/62d21710-5913-4131-9587-7264fd5c095a)

**Comment by GaryAustin1** [comment, marked as answer]
Update requires that you also meet the select policy as there is a where clause that needs to read the table to find the record.  You also have an additional select() that will require it.

Table 287 here spells it out:
https://www.postgresql.org/docs/current/sql-createpolicy.html

**Reply by Albert-Gao** [reply]
Thanks mate, does it mean, 

1. there is no way in this case I can make SELECT not work but update work?
I want to disable listing all profiles, but enable get a profile by id, also, enable update but only for a few fields, I suppose I have to create a EDGE function to wrap the admin sdk for such use case?

2. With such a policy, it disables CRUD against the profiles table?

```sql
create policy "profiles:disable read all" on profiles
    for select
    using (false);
```

**Reply by GaryAustin1** [reply]
You will have to use an rpc call to do the update so you can bypass select policy.  Make the function a security definer type, but make sure you enforce your own security as needed.
